fix(chain): fix block ordinal computation

bowenwang1996 opened the pull request:
In #4793 we unfortunately did not fix the issue fully. The situation that we did not fix is as follows: if block h is a fork and block h+1 is on the canonical chain, they would have the same ordinal and if block h is received later than block h+1, it would overwrite block h+1 in `ColBlockOrdinal`, which is incorrect.

This PR fixes the issue completely by updating `ColBlockOrdinal` in a similar way to how `ColBlockHeight` is updated. Essentially we only update it when the header head change. When the header head changes, we overwrite the ordinals between the last header height and the new header head. This way we guarantee that only hashes of blocks on the canonical chain eventually stay in the column given how the fork choice rule works.

Test plan
--------------------------
`test_block_merkle_proof`

mfornet commented on chain/chain/src/store.rs:
```diff
@@ -1745,12 +1741,13 @@ impl<'a> ChainStoreUpdate<'a> {
     }
 
+    pub fn save_block_ordinal(&mut self, block_hash: CryptoHash, block_ordinal: NumBlocks) {
+        self.chain_store_cache_update.block_ordinal_to_hash.insert(block_ordinal, block_hash);
```
I'm not very familiar with the code. Is this enough so that this change is ultimately reflected on the database?

bowenwang1996 replied:
Yes. It is saved when we call `ChainStoreUpdate::commit`

mfornet reviewed (commented):
If possible, let's try to verify every block/ordinal is properly computed and update all affected values.

bowenwang1996 commented:
@mfornet I added the DB migration. Also I changed it to a better implementation (essentially mimicking what we do to store the block height -> block hash mapping). This way we do not have to deal with some corner cases.

matklad commented on integration-tests/tests/client/process_blocks.rs:
```diff
@@ -1923,11 +1924,10 @@ fn test_incorrect_validator_key_produce_block() {
 }
 
-fn test_block_merkle_proof_with_len(n: NumBlocks) {
+fn test_block_merkle_proof_with_len(n: NumBlocks, rng: &mut StdRng) {
```
:+1:

matklad commented on integration-tests/tests/client/process_blocks.rs:
```diff
@@ -1949,8 +1954,23 @@ fn test_block_merkle_proof_with_len(n: NumBlocks) {
     let root = head.header().block_merkle_root();
+    // verify that block ordinal to block hash map is correct
+    for h in 0..head.header().height() {
+        if let Ok(block) = env.clients[0].chain.get_block_by_height(h).map(Clone::clone) {
```
```suggestion
        if let Ok(block) = env.clients[0].chain.get_block_by_height(h).cloned() {
```

bowenwang1996 replied:
This is a `Result`, so `cloned` doesn't work :(

matklad reviewed (approved):
Can/should we add an assert somewhere checking that we never overwrite the ordinal?

mfornet commented:
> Can/should we add an assert somewhere checking that we never overwrite the ordinal?

Sometimes it is ok to rewrite ordinals, if they were part of a fork. We must make sure not to overwrite ordinals from finalised blocks. 

The other option is not to write ordinals until the block becomes final, and then making sure they are not overwritten.

matklad commented:
>The other option is not to write ordinals until the block becomes final

I was under the impression that this is exactly what this PR intends to do? To quote Bowen, 

> This PR fixes the issue completely by only updating ColBlockOrdinal for finalized blocks.

bowenwang1996 commented:
@matklad sorry I forgot to update the description. Thanks for catching it!

mm-near commented on integration-tests/tests/client/process_blocks.rs:
```diff
@@ -1949,8 +1954,23 @@ fn test_block_merkle_proof_with_len(n: NumBlocks) {
+
     let head = blocks.pop().unwrap();
     let root = head.header().block_merkle_root();
+    // verify that block ordinal to block hash map is correct
```
what is "block hash map" ? I thought that we're comparing with ColBlockMerkleTree ?

bowenwang1996 replied:
it is [block ordinal to block hash] map. Will rephrase it to make it easier to parse

mm-near commented on nearcore/src/migrations.rs:
```diff
@@ -478,6 +479,28 @@ pub fn migrate_24_to_25(path: &Path) {
+/// Fix an issue with block ordinal (#5761)
+pub fn migrate_30_to_31(path: &Path, near_config: &NearConfig) {
+    let store = create_store(path);
+    if near_config.client_config.archive && &near_config.genesis.config.chain_id == "mainnet" {
```
The fact that we're running it "only on mainnet" - means that this code will not be executed in testnet etc.

What I'd do is:
 - remove this condition (so that code runs on testnet / betanet etc)
 - in 491 -> say that for "mainnet" we start from block 47443088, while for other networks, we go from block 1

bowenwang1996 replied:
> in 491 -> say that for "mainnet" we start from block 47443088, while for other networks, we go from block 1

I think this is debatable. Given that the check on mainnet (going from genesis block) took about 27 hours, I am not sure whether we want to do it on testnet as that will incur a significant amount of downtime for an archival node. At the same time, I manually tested this code on a mainnet archival node and it works fine (confirmed by Aurora).

@mfornet could you let us know whether it is important to run migration for testnet as well?

mfornet replied:
So far we haven't seen this issue on Testnet, but it is hard to know if the issue is present for some blocks in the past.
If it is the case it doesn't affect future txs, so from the POV of the bridge, running the migration in testnet is not necessary.

sept-en replied:
@bowenwang1996 Yes, we haven't experienced that issue on Testnet so far, therefore, I don't think this is required to run a migration there.

N.B.: I will let you know if we encounter this issue there.

mm-near commented on nearcore/src/migrations.rs:
```diff
@@ -478,6 +479,28 @@ pub fn migrate_24_to_25(path: &Path) {
+        for height in 47443088..=head.height {
+            if let Ok(block_hash) = chain_store.get_block_hash_by_height(height) {
+                let block_ordinal = chain_store.get_block_merkle_tree(&block_hash).unwrap().size();
+                store_update
```
could we read the previous value and see if the new value differs ?

So that we could print the total amount of changed rows in the end ? 

(while this is not stricktly needed - it would help with debugging - and we should see the correct result when we run this code on the canary node)

mm-near commented on nearcore/src/migrations.rs:
```diff
@@ -478,6 +479,28 @@ pub fn migrate_24_to_25(path: &Path) {
 }
 
+/// Fix an issue with block ordinal (#5761)
+pub fn migrate_30_to_31(path: &Path, near_config: &NearConfig) {
```
Maybe let's add the comment with "expected amount of time" such migration is going to take ?

bowenwang1996 replied:
This is difficult because it is a moving target. The time it took to migrate the node I used to test is not necessarily reflective of the time it will take to migrate a node when this change is released. Still, I added a note to document the amount of time it takes now.

nemani replied:
Is there a way to prevent migration if I am using https://github.com/aurora-is-near/partner-relayer-deploy/  and starting a new node from scratch ?

mm-near commented on nearcore/src/migrations.rs:
```diff
@@ -478,6 +479,28 @@ pub fn migrate_24_to_25(path: &Path) {
+            if let Ok(block_hash) = chain_store.get_block_hash_by_height(height) {
+                let block_ordinal = chain_store.get_block_merkle_tree(&block_hash).unwrap().size();
+                store_update
+                    .set_ser(DBCol::ColBlockOrdinal, &index_to_bytes(block_ordinal), &block_hash)
```
do we have any test case that is executing this code ?

bowenwang1996 replied:
tested it manually on a mainnet archival node

mm-near commented on chain/chain/src/store.rs:
```diff
@@ -3221,8 +3223,8 @@ mod tests {
             store_update.save_block(block.clone());
             store_update.inc_block_refcount(block.header().prev_hash()).unwrap();
-            store_update.save_head(&Tip::from_header(block.header())).unwrap();
             store_update.save_block_header(block.header().clone()).unwrap();
```
so the ordering here is important ? was it a bug before ?

mm-near replied:
should we add something (in separate PR) to enforce the proper ordering ?

bowenwang1996 replied:
this is a piece of testing code

mm-near commented on nearcore/src/migrations.rs:
```diff
@@ -478,6 +479,28 @@ pub fn migrate_24_to_25(path: &Path) {
+        let mut chain_store = ChainStore::new(store.clone(), genesis_height);
+        let head = chain_store.head().unwrap();
+        let mut store_update = BatchedStoreUpdate::new(&store, 10_000_000);
+        // we manually checked mainnet archival data and the first block where the discrepancy happened is `47443088`.
```
did we verify every archival node that exists ? (as AFAIK this bug was not deterministic)

bowenwang1996 replied:
We verified it on multiple of our archival nodes and they all have the same issue.

bowenwang1996 commented:
Update: I ran the PR on a mainnet archival node. The migration was done successfully and the Aurora team confirmed that the fix works well.
